### ISSUE TITLE
lucene timing probably not comparable to single-threaded timing

### ISSUE THREAD

**Issue by jjfiv** [lintool/IR-Reproducibility#15]
Lucene's timing numbers are unfair: IndexSearcher uses Future<T> to compute in parallel on shards by default.

https://github.com/apache/lucene-solr/blob/lucene_solr_5_2_1/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L706

**Comment by chatman** [comment]
I think here, the search is performed on various Lucene segments in parallel. As documents keep getting added, the commits ensure that segements are merged to form bigger segments automatically. 
I don't remember what was the number of segments by the end of indexing. If only single thread is a hard limitation for a benchmark, we could just merge all segments into one and rerun the evaluation.

**Comment by vigna** [comment]
> On 7 Oct 2015, at 23:22, Ishan Chattopadhyaya notifications@github.com wrote:
> 
> I don't remember what was the number of segments by the end of indexing. If only single thread is a hard limitation for a benchmark, we could just merge all segments into one and rerun the evaluation.

Probably this is a bit complicated to get in the paper, but at least to have a gut feeling of what's the magnitude of the improvement (if there's one) would be nice.

Ciao,

```
                seba
```

**Comment by iorixxx** [comment]
TREC collections are absolutely static. `writer.forceMerge(1)` would make searches even faster, no?

Lucene search times are tricky, Operating System disk caches have huge impact on search times.
It may require to flush/empty OS disk caches before every search.

**Comment by jjfiv** [comment]
The answer is always maybe until you have some numbers. `writer.forceMerge(1)` is touted as making search faster in the tutorial. The can of worms I haven't opened is that if it does and we want to present that better time, it means that Lucene's indexing time is wrong, as the tutorial also claims that `writer.forceMerge(1)` is quite time-expensive.

Your point about disk caches making timing tricky stands, but in a world where the whole collection fits in ram, and there are 31 other bored cores, and each document score is independent, I'd expect to see a large benefit from threads.

Since there was so much RAM on the EC2 instance I'd be surprised if disk contention was a problem. I'm sure there's a diminishing returns rule here that depends on the communication needed for the multithreaded approach -- it may even help some queries and hurt others.

I'd be happy and impressed if Lucene was faster than most systems on a single-thread, but we don't yet know that. It's probably true, but we should be careful about making a claim that we're comparing systems if they're working with vastly different resources: a single-threaded scorer vs. a multithreaded one, particularly on such an embarrassingly parallel problem like doc-at-a-time scoring.

**Comment by vigna** [comment]
> On 8 Oct 2015, at 16:58, John Foley notifications@github.com wrote:
> 
> The answer is always maybe until you have some numbers. writer.forceMerge(1) is touted as making search faster in the tutorial. The can of worms I haven't opened is that if it does and we want to present that better time, it means that Lucene's indexing time is wrong, as the tutorial also claims that writer.forceMerge(1) is quite time-expensive.
> 
> Your point about disk caches making timing tricky stands, but in a world where the whole collection fits in ram, and there are 31 other bored cores, and each document score is independent, I'd expect to see a large benefit from threads.

Well, the point is to gather some knowledge. Maybe it's better, maybe it's worse, maybe forceMerge() takes forever. If nobody tries, we will never know :).

Ciao,

```
                seba
```

**Comment by iorixxx** [comment]
> maybe forceMerge() takes forever. 

It took 40 minutes to optimize a ClueWeb09B (90 GB) index for me.

Here is a quote from gov2.md:

> The average is taken from three runs of the queries. 

Second and the third run will be instantaneous for Apache Lucene.

**Comment by vigna** [comment]
> On 8 Oct 2015, at 19:01, Ahmet Arslan notifications@github.com wrote:
> 
> maybe forceMerge() takes forever.
> 
> It took 40 minutes to optimize a ClueWeb09B (90 GB) index for me.

OK, so it shouldn't impact to much on the timings for GOV2.

> Here is a quote from gov2.md:
> 
> The average is taken from three runs of the queries.
> 
> Second and the third run will be instantaneous for Apache Lucene.

This is always an issue with Lucene--also in other context, doing proper measurement is difficult because of caching. Is there any simple way to disable it?

Ciao,

```
                seba
```

**Comment by iorixxx** [comment]
> Is there any simple way to disable it

"clear your disk caches or reboot the machine between the tests"
http://blog.thetaphi.de/2012/07/use-lucenes-mmapdirectory-on-64bit.html?showComment=1419514694910#c3669572939426142499

**Comment by vigna** [comment]
> On 8 Oct 2015, at 19:22, Ahmet Arslan notifications@github.com wrote:
> 
> Is there any simple way to disable it
> 
> "clear your disk caches or reboot the machine between the tests"
> http://blog.thetaphi.de/2012/07/use-lucenes-mmapdirectory-on-64bit.html?showComment=1419514694910#c3669572939426142499

I'm a bit confused.

I thought we were talking about an _internal_ caching mechanism (I know there's one). All reasonable indices will use in-memory or memory-mapped indices, and all will use the same caching.

Just to know: Jimmy, is it possible to have a file with the timing of the three measurement for all systems? This could easily settle the issue.

Ciao,

```
                seba
```

**Comment by snapbug** [comment]
The raw log files that all the data was extracted from are in the [-exp](https://github.com/lintool/IR-Reproducibility-exp) repository. In particular to this, the [extracted timings](https://github.com/lintool/IR-Reproducibility-exp/blob/master/allms).

There's not really much change between runs, the timing for an individual query are essentially unchanging.

For Lucene's BM25 Count run:
![lucene-count](https://cloud.githubusercontent.com/assets/945966/10384541/30d74848-6e99-11e5-8c27-4380f57f16ed.png)

This is the same behaviour as seen by ATIRE (BM25 run):
![atire-bm25](https://cloud.githubusercontent.com/assets/945966/10384540/30923f0a-6e99-11e5-95af-59fa1343ab85.png)

**Comment by vigna** [comment]
> On 9 Oct 2015, at 04:23, Matt Crane notifications@github.com wrote:
> 
> The raw log files that all the data was extracted from are in the -exp repository. In particular to this, the extracted timings.

I'd say disk caching is a non-problem. Maybe the instance worked on an SSD. Go figure.

Ciao,

```
                seba
```

**Comment by iorixxx** [comment]
Why record running times on a per-query basis?
Here is my execution times in Second:MilliSecond format for the ClueWeb09B dataset.
I don't have SSD. As can be observed from the table, the difference between the first run and the subsequent runs is notable for me. When I re-run the same program to the produce timings given in table, the difference disappears.

I will provide the benchmark code shortly.

| Topic_Set | i=1 | i=2 | i=3 |
| --- | --- | --- | --- |
| topics.web.1-50 | 45:50 | 5:52 | 5:94 |
| topics.web.51-100 | 36:86 | 6:77 | 6:09 |
| topics.web.101-150 | 32:32 | 3:39 | 3:24 |
| topics.web.151-200 | 21:01 | 3:82 | 3:42 |

**Comment by chatman** [comment]
Indeed, the correct way to do the re-runs would be to run the program again. I think this is what the script is doing currently. Running the same query more than once from the same process (program invocation) doesn't make sense.

**Comment by iorixxx** [comment]
Here is the [program](https://github.com/iorixxx/Anserini/blob/master/src/main/java/io/anserini/SearchTimeUtil.java) that is used to produce the numbers in the tables. [README](https://github.com/iorixxx/Anserini/blob/master/README.md) has documentation regarding how to use it.

> re-runs would be to run the program again.

Even if you re-run, OS disk caches are involved. Here is the same benchmark results after a re-run. Difference is gone between first and third.

```
topics.web.1-50_1 search completed in 00:05:662
topics.web.1-50_2 search completed in 00:04:917
topics.web.1-50_3 search completed in 00:04:928
topics.web.51-100_1 search completed in 00:06:380
topics.web.51-100_2 search completed in 00:06:218
topics.web.51-100_3 search completed in 00:06:184
topics.web.101-150_1 search completed in 00:03:184
topics.web.101-150_2 search completed in 00:03:020
topics.web.101-150_3 search completed in 00:03:050
topics.web.151-200_1 search completed in 00:03:192
topics.web.151-200_2 search completed in 00:03:109
topics.web.151-200_3 search completed in 00:03:105
```

**Comment by gsingers** [comment]
I am curious, how do other systems do this out of the box?  I haven't looked much at Terrier, Galago and the like.  @jjfiv are you saying their default setup is single threaded?  Query time on a single optimized segment in Lucene has always been (on average) faster than what is the current default, it's just rarely done in practice due to the cost of optimization and very few people have static indices these days.

**Comment by chatman** [comment]
If someone gets a chance, please review the following commit for this issue:
https://github.com/lintool/IR-Reproducibility/commit/3968d5841019e103bb26430b44b774a6915133a5
In the hurry to get this done, I made the script slightly more ugly. I intend to clean this up a bit more if I get a chance to do so, but not a priority.

For the counts index, the force merge took about 6m2s on top of the indexing time. For the query times, there was a ~10-15% speedup. Thanks @jjfiv for bringing this up.
